Feat/load image

HLoc26 opened the pull request:
# This branch is built on top of feat/search (#51)

### Pull Request Description: Implement `loadImage` Feature for Room Images

#### Summary
This PR introduces the `loadImage` feature, enabling the application to dynamically load and display room images from a static server. The feature ensures that default images are displayed when specific images are unavailable, improving the user experience.

#### Changes Made
1. **Static Server Setup**:
   - Configured an Nginx-based static server to serve image files from the `/uploads` directory.
   - Added a custom nginx.conf file to define the `/uploads/` route and enable file listing for debugging purposes.
   - Created a Dockerfile for the static server to copy image assets and configuration.

2. **Default Images**:
   - Added default images for rooms and accommodations:
     - `uploads/room/default.jpg`
     - `uploads/accommodation/default.jpg`
   - Ensures that default images are displayed when specific images are missing.

3. **Frontend Integration**:
   - Updated the `RoomCard` component to use the `loadImage` feature:
     - Dynamically loads images from the static server using the `/uploads/` route.
     - Displays the default image (`/uploads/room/default.jpg`) if no images are provided for a room.

4. **Proxy Configuration**:
   - Updated the Vite development server configuration (`vite.config.js`) to proxy requests to the static server:
     - Proxy `/uploads` requests to the static server running on port `3001`.

5. **Docker Integration**:
   - Updated docker-compose.yml to include the static service:
     - Maps the `uploads` directory to `/usr/share/nginx/html/uploads` in the container.
     - Ensures the static server is accessible on port `3001`.

#### How to Test
1. **Static Server**:
   - Start the static server using `docker-compose up`.
   - Access the static server at `http://localhost:3001/uploads/` to verify file availability.

2. **Frontend**:
   - Navigate to the room listing page.
   - Verify that rooms with no images display the default image (`/uploads/room/default.jpg`).
   - Verify that rooms with images load the correct images from the static server.

3. **Edge Cases**:
   - Test with missing images to ensure the default image is displayed.
   - Test with various image formats (e.g., `.jpg`, `.png`) to ensure compatibility.

#### Related Files
- Dockerfile
- nginx.conf
- default.jpg
- default.jpg
- RoomCard.jsx
- vite.config.js
- docker-compose.yml

HLoc26 committed:
feat(static): add Nginx configuration and Docker setup for static file serving

HLoc26 committed:
feat(vite): add proxy configuration for uploads

HLoc26 committed:
feat(HotelCard, AccommodationDetail): add thumbnail support and default image fallback

HLoc26 committed:
refactor(Search): streamline search results fetching logic

HLoc26 committed:
feat(Search): enhance date formatting in search parameters

HLoc26 committed:
feat(RoomCard): add default image fallback for room images

HLoc26 committed:
Merge branch 'develop' into feat/loadImage

Bright-04 commented on UI/src/components/ui/RoomCard/RoomCard.jsx:
```diff
@@ -0,0 +1,255 @@
+	const [currentImageIndex, setCurrentImageIndex] = React.useState(0);
+	const [modalOpen, setModalOpen] = React.useState(false);
+
+	const imagesList = images ? images : [{ img: "/uploads/room/default.jpg" }];
```
In RoomCard.jsx, you're using this pattern:

Bright-04 commented on UI/src/pages/AccommodationDetail/AccommodationDetail.jsx:
```diff
@@ -0,0 +1,250 @@
+					<Box mt={3}>
+						{/* Main image */}
+						<Box>
+							<img
+								src={images ? images[selectedImage].filename : "/uploads/accommodation/default.jpg"}
+								alt="Accommodation"
+								style={{ width: "100%", maxHeight: 500, objectFit: "cover", borderRadius: 8 }}
+							/>
```
While in AccommodationDetail.jsx, a different approach is used:

Bright-04 commented on UI/src/components/ui/RoomCard/RoomDetailModal.jsx:
```diff
@@ -0,0 +1,228 @@
+								objectFit: "cover",
+							}}
+						>
+							{/* Current Image */}
+							{images?.length > 0 ? (
+								<img
+									srcSet={`${images[currentImageIndex].img}?w=800&h=600&fit=cover&auto=format&dpr=2 2x`}
+									src={`${images[currentImageIndex].img}?w=800&h=600&fit=cover&auto=format`}
+									alt={images[currentImageIndex].title}
+									loading="lazy"
+									style={{
+										width: "100%",
+										height: "100%",
+										objectFit: "cover",
+									}}
+								/>
+							) : (
+								<Typography>No images available</Typography>
+							)}
```
And in RoomDetailModal.jsx, yet another pattern:

Bright-04 reviewed (commented):
I can see that you implemented multiple fallback patterns, which may cause major inconsistency.

HLoc26 committed:
fix(RoomCard, RoomDetailModal, AccommodationDetail): correct image property references for consistency

Bright-04 commented on UI/src/components/ui/RoomCard/RoomCard.jsx:
```diff
@@ -0,0 +1,254 @@
+							}}
+						>
+							{/* Current Image */}
+							{imagesList?.length > 0 ? (
+								<img
+									srcSet={`/uploads/room/${imagesList[0].filename}`}
+									src={`/uploads/room/${imagesList[0].filename}`}
+									loading="lazy"
+									style={{
+										width: "100%",
+										height: "100%",
+										objectFit: "cover",
+									}}
+								/>
+							) : (
+								<Typography>No images available</Typography>
+							)}
```
The code doesn't properly handle image loading failures. If 'imagesList[0].filename' exists but the actual image file doesn't exist on the server, it will show a broken image icon instead of falling back to the default image.

sang-ute commented:
New commits `[5d98f28](https://github.com/HLoc26/VinaBooking/pull/52/commits/5d98f28f306d6a8067b42fb26f5d43a3d6c20001)` seems to make the room image disabled

Bright-04 committed:
fix(RoomCard): update default image path for consistency in image handling

Bright-04 commented:
## 1. Inconsistent Image Path Construction

Your code has several inconsistent approaches to constructing image paths:

```jsx
// RoomCard.jsx - Direct path construction
<img
  srcSet={`/uploads/room/${imagesList[0].filename}`}
  src={`/uploads/room/${imagesList[0].filename}`}
  loading="lazy"
  // No error handling
/>

// RoomDetailModal.jsx - Similar but in a different component
<img
  srcSet={`/uploads/room/${images[currentImageIndex].filename}`}
  src={`/uploads/room/${images[currentImageIndex].filename}`}
  alt={images[currentImageIndex].title}
  // Different approach to the same problem
/>

// AccommodationDetail.jsx - Another variant
<img
  src={`/uploads/accommodation/${images ? images[selectedImage].filename : "default.jpg"}`}
  alt="Accommodation"
  // Inline fallback logic duplicated
/>
```

Which leads to...
- Risk of errors when image storage paths change
- Different behaviors for fallback paths
- Maintenance nightmare when updating path patterns

## 2. Duplicated Image Loading Logic

Each component implements its own version:

```jsx
// In RoomCard.jsx - Manual fallback
const imagesList = images && images.length > 0 ? images : [{ filename: "default.jpg" }];

// In AccommodationDetail.jsx - Inline ternary
src={`/uploads/accommodation/${images ? images[selectedImage].filename : "default.jpg"}`}

// In ReviewCard.jsx - Different loading state implementation
const [loading, setLoading] = useState(true);
const handleImageLoad = useCallback(() => {
  setLoading(false);
}, []);

// Missing in many components - No consistent error handling
const handleImageError = () => {
  setError(true);
  setImageLoading(false);
};
```

Which leads to...
- Inconsistent user experience (some components handle errors, others don't)
- Higher maintenance burden (bug fixes need to be applied in multiple places)
- Different loading indicators across components

## 3. No Centralized Image Service/Utility

The codebase lacks a single source of truth for image-related functionality:

```jsx
// These functionalities are reimplemented in each component rather than centralized:

// Path construction
`/uploads/${type}/${filename}`

// Default fallback
images && images.length > 0 ? images : [{ filename: "default.jpg" }];

// Loading states
const [loading, setLoading] = useState(true);

// Error handling
const handleImageError = () => { /* different in each component */ };
```

> Idk bro, currently everything works fine, but it's a mess that needs to think about...

HLoc26 committed:
feat(AppImage, imageHelper): add AppImage component and utility for image URL generation

HLoc26 committed:
feat(useSafeImageList): implement hook for safe image list handling

HLoc26 committed:
feat(RoomCard, RoomDetailModal, AccommodationDetail): integrate AppImage component and useSafeImageList hook for improved image handling

HLoc26 committed:
fix: update import paths for useSafeImageList hook in RoomCard, RoomDetailModal, and AccommodationDetail components

HLoc26 committed:
feat: Appluing AppImage and useSafeImageList in ReviewCard

Bright-04 committed:
feat(ReviewCard): add AppImage component

feat(ReviewCard): replace img tag with AppImage component for improved image handling

HLoc26 committed:
Merge branch 'develop' into feat/loadImage
